### PR TITLE
[Telemetry] Add support to collect the Static Node Count from Blueprints

### PR DESCRIPTION
**Pull request by kadupoornima**
**Objective**
This PR introduces telemetry collection for the static node counts defined in a blueprint, addressing the static count requirements. It extracts both top-level and inline explicit node counts mapped by their respective machine types. These changes provide better insights into the scale and composition of deployed infrastructure without requiring significant manual overhead.

**Implementation Details**
The PR introduces a new telemetry key, `CLUSTER_TOOLKIT_STATIC_NODE_COUNTS`. It targets specific configuration settings (`static_node_count`, `node_count_static`, and `instance_count`) to compute these values. The output is formatted as a comma-separated string mapping of machine types to counts, trimmed of JSON curly braces (e.g., `"g4-standard-48":3,"a3-ultragpu-8g":2`).

**Changes:**

* **Telemetry Enhancement**: Introduced the collection of static node counts as a new telemetry metric to improve infrastructure visibility.
* **Recursive Extraction Logic**: Implemented robust helper functions to recursively traverse blueprint structures, enabling the aggregation of node counts from complex configurations like Slurm V6 partitions.
* **Code Refactoring**: Refactored machine type extraction logic into reusable helper functions to improve maintainability and reduce code duplication.
* **Comprehensive Testing**: Added extensive unit tests to verify accurate node count aggregation across various module types and edge cases.

### CODE REVIEW

File: pkg/telemetry/collector.go [hunk 1]
```diff
@@ -35,6 +36,12 @@ var (
 		"node_type",                     // For modules that use node_type setting instead of machine_type to set machines.
 		"system_node_pool_machine_type", // For gke-cluster system node pools.
 	}
+	staticNodeCountSettings = []string{
+		"static_node_count",                   // Used in GKE node pool. If set, autoscaling will be disabled. Defaults to 0.
+		"node_count_static",                   // Standalone Slurm V6 CPU and TPU nodesets use 'node_count_static'. Defaults to 0.
+		"instance_count",                      // VM instances and Batch login nodes use 'instance_count' to define static nodes. Default is 1.
+		"nodeset", "nodeset_tpu", "partition", // Combine top-level explicit keys and complex inline object list keys for Slurm V6.
+	}
```
[the remark this answers is not in the document]
Comment (kadupoornima) [reply pkg/telemetry/collector.go:45]: Outdated.

File: pkg/telemetry/collector_util.go [hunk 2]
```diff
@@ -223,6 +231,146 @@ func extractDefaultMachineType(key string, m config.Module) string {
...
+		for _, input := range res.mi.Inputs {
+			if input.Name == key && input.Default != nil {
+				if val, ok := input.Default.(T); ok {
+					return val
+				}
+			}
+		}
+	case <-time.After(500 * time.Millisecond):
+	}
+
```
[the remark this answers is not in the document]
Comment (kadupoornima) [reply pkg/telemetry/collector_util.go:415]: Outdated.

File: pkg/telemetry/collector.go [hunk 3]
```diff
@@ -276,6 +269,27 @@ func getModules(modulesList []string) string {
...
+		}
+	}
+
+	counts, err := json.Marshal(countsByMachineType)
+	if err != nil || len(countsByMachineType) == 0 {
+		return ""
+	}
+
+	// Trimming the curly braces safely extracts our exact format `"g4-standard-48":3,"a3-ultragpu-8g":2`
+	return strings.Trim(string(counts), "{}")
```
[the remark this answers is not in the document]
Comment (kadupoornima) [reply pkg/telemetry/collector.go:293]: Ack.

File: pkg/telemetry/collector_util.go [hunk 4]
```diff
@@ -206,21 +326,56 @@ func extractDefaultMachineType(key string, m config.Module) string {
...
+		case float64:
+			// Safely cast float64 to int if T is int
+			if _, isInt := any(zero).(int); isInt {
+				return any(int(v)).(T), true
+			}
+		case float32:
+			if _, isInt := any(zero).(int); isInt {
+				return any(int(v)).(T), true
+			}
+		}
```
[the remark this answers is not in the document]
Comment (kadupoornima) [reply pkg/telemetry/collector_util.go:454]: Done.

File: pkg/telemetry/collector_util.go [hunk 5]
```diff
@@ -167,37 +176,148 @@ func extractExplicitMachineType(bp config.Blueprint, key string, m config.Module
...
+}
+
+// processInlineItem extracts machine types and counts from an individual map/object in an inline list.
+func processInlineItem(item cty.Value, topMachineType string, counts map[string]int) {
+	item, _ = item.Unmark()
+	ty := item.Type()
+
+	if !ty.IsObjectType() && !ty.IsMapType() {
+		return
+	}
```
[the remark this answers is not in the document]
Comment (kadupoornima) [reply pkg/telemetry/collector_util.go:244]: Done.

File: pkg/telemetry/collector_util.go [hunk 6]
```diff
@@ -167,60 +176,283 @@ func extractExplicitMachineType(bp config.Blueprint, key string, m config.Module
...
+	if !hasTopology {
+		return 0, false
+	}
+
+	if m.Settings.Has("enable_flex_start") {
+		val, err := bp.Eval(m.Settings.Get("enable_flex_start"))
+		if err == nil && val.Type() == cty.Bool && !val.IsNull() && val.IsKnown() && val.True() {
+			return 0, false
+		}
+	}
```
[the remark this answers is not in the document]
Comment (kadupoornima) [reply pkg/telemetry/collector_util.go:325]: Done.

File: pkg/telemetry/collector_util.go [hunk 7]
```diff
@@ -167,60 +176,283 @@ func extractExplicitMachineType(bp config.Blueprint, key string, m config.Module
...
+			return val
+		}
+	}
+	if val, ok := extractDefaultSetting[int]([]string{key}, m); ok {
+		if val > 0 {
+			return val
+		}
+	}
+	return 1
+}
```
[the remark this answers is not in the document]
Comment (kadupoornima) [reply pkg/telemetry/collector_util.go:307]: Done.

File: pkg/config/hardware.go [hunk 8]
```diff
@@ -149,7 +149,7 @@ func evalString(bp Blueprint, val cty.Value) (string, bool) {
 	return "", false
 }
```
Comment (SwarnaBharathiMantena) [inline pkg/config/hardware.go:151]: Lets add a comment for this exported function: "// ExtractTopology extracts the TPU topology string from the given module settings or placement policy."

File: pkg/telemetry/collector_util.go [hunk 9]
```diff
@@ -167,60 +176,284 @@ func extractExplicitMachineType(bp config.Blueprint, key string, m config.Module
...
+	}
+
+	pools := getMultiplier("num_node_pools", m, bp)
+	slices := getMultiplier("num_slices", m, bp)
+	multiplier := max(slices, pools)
+
+	return baseCount * multiplier, true
+}
+
+func getMultiplier(key string, m config.Module, bp config.Blueprint) int {
```
Comment (SwarnaBharathiMantena) [inline pkg/telemetry/collector_util.go:294]: Suggestion on code here - 

Fetch the `ModuleInfo` once inside `getModuleNodeCounts` (or inside `getTopLevelNodeCount`) and pass it down as an optional argument to avoid redundant slow network or disk operations:

```
// Pass optional ModuleInfo to avoid redundant parsing/cloning calls
func getMultiplier(key string, m config.Module, bp config.Blueprint, mi *modulereader.ModuleInfo) int {
	if val, ok := extractExplicitIntSetting(key, m, bp); ok {
		if val > 0 {
			return val
		}
		return 1
	}
	if mi != nil {
		for _, input := range mi.Inputs {
			if val, ok := parseDefaultValue[int](input.Name, input.Default, key); ok {
				if val > 0 {
					return val
				}
			}
		}
	}
	return 1
}
```

File: pkg/config/hardware.go (hunk 8 above)
Comment (kadupoornima) [reply pkg/config/hardware.go:151]: Done.

File: pkg/telemetry/collector_util.go (hunk 9 above)
Comment (kadupoornima) [reply pkg/telemetry/collector_util.go:294]: Ack. There is a 500ms timeout on these calls anyway.

**Review by SwarnaBharathiMantena — APPROVED** [review]
LGTM!
